sys/include/storage/ftl.h

```diff
@@ -0,0 +1,465 @@
+    if(ret != 0) { Error }
+
+    assert(header.data_length == 503);
+
+To know in advance how much data can be written per subpage, one can use the
+#ftl_data_per_subpage function:
+
+    int data_length = ftl_data_per_subpage(&device, false);
+    int data_length_with_ecc = ftl_data_per_subpage(&device, true);
+
+    assert(data_length == 509);
+    assert(data_length_with_ecc == 503);
+
+ECC
+===
+
+TODO: write about ecc
+
+ *
+ *
+ * @{
+ * @author      Lucas Jenß <lucas@x3ro.de>
+ */
+
+#ifndef _FS_FTL_H
+#define _FS_FTL_H
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @name Data Types
+ * @{
+ */
+
+// This is needed here since `ftl_partition_s` and `ftl_device_s` reference each other
+struct ftl_device_s;
+
+/**
+ * @brief A partition on a device managed by the FTL.
+ */
+typedef struct ftl_partition_s {
+    struct ftl_device_s *device;    //!< The device on which the partition is located
+    uint32_t base_offset;           //!< Zero-indexed absolute offset of the partition __in blocks__
+    uint32_t size;                  //!< Size of the partition __in blocks__
+
+    /**
+     * The next page to be written by the FTL, or in other words,
+     * the first known free page
+     */
+    uint32_t next_subpage;
+
+    /**
+     * Page that was last written in this partition. Must not necessarily be next_page-1,
+     * because the previous block may have been faulty
+     */
+    uint32_t last_written_subpage;
+
+    uint32_t erased_until;          //!< Pointer to the last erased block
+    uint32_t free_until;            //!< Pointer to the last known free block
+} ftl_partition_s;
+
+/**
+ * @brief A device managed by the FTL.
+ */
+typedef struct ftl_device_s {
+    uint32_t metadata_version;
+
+    uint32_t total_pages;     //!< Total amount of pages configured for the device
+    uint16_t page_size;       //!< Page size configured for the device
+    uint16_t subpage_size;    //!< Size of a single subpage in bytes
+    uint16_t pages_per_block; //!< Amount of pages inside an erase segment (block)
+    uint8_t ecc_size;         //!< Size of the ECC determined for device's subpage size
+    bool is_initialized;      //!< Whether #ftl_init was successfully called
+
+    uint8_t partition_count;      //!< Number of partitions allocated for this device
+    ftl_partition_s **partitions; //!< Array of partitions allocated for this device
+
+    unsigned char *_subpage_buffer; //!< Buffer for subpage read/write operations.
+    unsigned char *_ecc_buffer;     //!< Buffer for ECC calculation
+
+    /**
+     * Callback which must write a data buffer of the given length to a certain offset
+     * inside a page.
+     */
+    int (*_write)(const unsigned char *buffer,
+                              uint32_t page,
+                              uint32_t offset,
+                              uint16_t length);
```

Comment:
not officially a standard thing (yet), but most APIs follow this order, and (maybe personal opinion) I think it is nicer to keep parameters closely coupled next to each other (buf and len in this case.

But for changing the buffer type, this is AFAIK RIOT standard -> the same was done e.g. for all periph interfaces lately...